### PR TITLE
fix: Support the verbose output while running the install and upgrade command

### LINKED ISSUE #44: Support  the verbose output while running the install and upgrade command
# Background

To run Coffe you are run the following command `cargo run -- --help` and you can run the following command to setup coffee for your core lightning node

```
cargo run -- setup <path of the core lightning file>
cargo run -- remote add lightningd https://github.com/lightningd/plugins.git
cargo run -- install <plugin_name>
```

Currently when you are running `install` you see in the terminal the install output like the python output, this is not good if we want to implement a Terminal UI https://github.com/coffee-tools/coffee/issues/16

The goal of this task is to hide the output of the install command and show the output caused by the python command just when the user specifies it.

# Goal

Be able to run the following command 

```
cargo run -- install <plugin_name> --verbose
```

### PR DESCRIPTION
**Pull request by Harshit933**
In the main.rs file, 

Running the install command without the --verbose flag, produces this ouput:
{Note: the file is already present on my local system that's why it is showing the error}

```
$ cargo run -- install rebalance 
 Finished dev [unoptimized + debuginfo] target(s) in 0.09s
     Running `target/debug/coffee_cmd install rebalance`
[2023-02-28T08:05:21Z ERROR coffee_lib::errors] ERROR #1: field plugin with value /home/harshit/.coffee/repositories/lightningd/rebalance/rebalance.py already present
thread 'main' panicked at 'code: 1, msg: field plugin with value /home/harshit/.coffee/repositories/lightningd/rebalance/rebalance.py already present', coffee_cmd/src/main.rs:42:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Running the install command with the --verbose flag, produces this ouput:

```
cargo run -- install rebalance --verbose
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
     Running target/debug/coffee_cmd install rebalance --verbose
Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: pyln-client>=0.12 in /home/harshit/.local/lib/python3.10/site-packages (from -r /home/harshit/.coffee/repositories/lightningd/rebalance/requirements.txt (line 1)) (0.12.1)
Requirement already satisfied: pyln-bolt7>=1.0 in /home/harshit/.local/lib/python3.10/site-packages (from pyln-client>=0.12->-r /home/harshit/.coffee/repositories/lightningd/rebalance/requirements.txt (line 1)) (1.0.246)
Requirement already satisfied: pyln-proto>=0.12 in /home/harshit/.local/lib/python3.10/site-packages (from pyln-client>=0.12->-r /home/harshit/.coffee/repositories/lightningd/rebalance/requirements.txt (line 1)) (0.12.0)
Requirement already satisfied: PySocks<2.0.0,>=1.7.1 in /home/harshit/.local/lib/python3.10/site-packages (from pyln-proto>=0.12->pyln-client>=0.12->-r /home/harshit/.coffee/repositories/lightningd/rebalance/requirements.txt (line 1)) (1.7.1)
Requirement already satisfied: cryptography<37.0.0,>=36.0.1 in /home/harshit/.local/lib/python3.10/site-packages (from pyln-proto>=0.12->pyln-client>=0.12->-r /home/harshit/.coffee/repositories/lightningd/rebalance/requirements.txt (line 1)) (36.0.2)
Requirement already satisfied: base58<3.0.0,>=2.1.1 in /home/harshit/.local/lib/python3.10/site-packages (from pyln-proto>=0.12->pyln-client>=0.12->-r /home/harshit/.coffee/repositories/lightningd/rebalance/requirements.txt (line 1)) (2.1.1)
Requirement already satisfied: coincurve<18.0.0,>=17.0.0 in /home/harshit/.local/lib/python3.10/site-packages (from pyln-...
```

shows all the tasks running behind the process.

In the `coffee_lib/src/plugins.rs` file I've just checked whether the --verbose flag is present or not.

fixes #44

### CODE REVIEW

**Comment by Harshit933** [comment]
fixes #44

File: coffee_cmd/src/coffee/cmd.rs [hunk 1]
```diff
@@ -19,7 +19,12 @@ pub struct CoffeeArgs {
 pub enum CoffeeCommand {
     /// Install a single by name.
     #[clap(arg_required_else_help = true)]
-    Install { plugin: String },
+    Install {
+         plugin: String, 
+
```
Comment (vincenzopalazzo) [inline coffee_cmd/src/coffee/cmd.rs:24]: no need line space

File: coffee_cmd/src/coffee/mod.rs [hunk 2]
```diff
@@ -18,6 +18,7 @@ use std::vec::Vec;
 pub mod cmd;
 mod config;
 
+
```
Comment (vincenzopalazzo) [inline coffee_cmd/src/coffee/mod.rs:21]: same there

File: coffee_cmd/src/main.rs [hunk 3]
```diff
@@ -1,5 +1,7 @@
 mod coffee;
 
+// use std::ptr::null;
```
Comment (vincenzopalazzo) [inline coffee_cmd/src/main.rs:3]:
```suggestion
```

File: coffee_lib/src/plugin.rs [hunk 4]
```diff
@@ -5,8 +5,12 @@ use serde::{Deserialize, Serialize};
 use std::fmt;
 use tokio::process::Command;
 
+
```
Comment (vincenzopalazzo) [inline coffee_lib/src/plugin.rs:8]: space no needed

File: coffee_lib/src/plugin.rs [hunk 5]
```diff
@@ -5,8 +5,12 @@ use serde::{Deserialize, Serialize};
...
 /// Plugin language definition
 #[derive(Clone, Debug, Serialize, Deserialize)]
+
+
+
```
Comment (vincenzopalazzo) [inline coffee_lib/src/plugin.rs:13]: space no needed

File: coffee_lib/src/plugin.rs [hunk 6]
```diff
@@ -18,21 +22,34 @@ pub enum PluginLang {
...
                     .arg("install")
+                    .arg("-q")
                     .arg("-r")
                     .arg(req_file.as_str())
                     .spawn()
                     .expect("not possible run the command");
-                let _ = child.wait().await?;
+                    let _ = child.wait().await?;
+                }
+
```
Comment (vincenzopalazzo) [inline coffee_lib/src/plugin.rs:52]: This works but now you are duplicating code, and also you do not prevent tokio to write to the std in this way.

The solution that I was in mind is something like that, when verbose is specified you do not use `spawn` but use output https://docs.rs/tokio/latest/tokio/process/struct.Command.html#method.output

```diff
+++ b/coffee_lib/src/plugin.rs
@@ -25,14 +25,17 @@ impl PluginLang {
                  * 2. return the path of the main file */
                 let req_file = format!("{path}/requirements.txt");
                 let main_file = format!("{path}/{name}.py");
-                // FIXME: enable the verbose command
-                let mut child = Command::new("pip")
-                    .arg("install")
-                    .arg("-r")
-                    .arg(req_file.as_str())
-                    .spawn()
-                    .expect("not possible run the command");
-                let _ = child.wait().await?;
+                let mut cmd = Command::new("pip");
+                cmd.arg("install").arg("-r").arg(&req_file.clone());
+                if verbose {
+                    let _ = cmd
+                        .spawn()
+                        .expect("Unable to run the command")
+                        .wait()
+                        .await?;
+                } else {
+                    let _ = cmd.output().await?;
+                };
                 Ok(main_file)
             }
```

In this way also future debug is more simple

What do you think?

File: coffee_lib/src/plugin.rs [hunk 7]
```diff
@@ -94,7 +111,7 @@ impl Plugin {
     /// configure the plugin in order to work with cln.
     ///
     /// In case of success return the path of the executable.
-    pub async fn configure(&mut self) -> Result<String, CoffeeError> {
+    pub async fn configure(&mut self, verbose: &bool) -> Result<String, CoffeeError> {
```
Comment (vincenzopalazzo) [inline coffee_lib/src/plugin.rs:114]:
```suggestion
    pub async fn configure(&mut self, verbose: bool) -> Result<String, CoffeeError> {
```

Yo do not need to borrow a bool :) it is fast to copy it

File: coffee_lib/src/plugin.rs [hunk 8]
```diff
@@ -107,10 +124,10 @@ impl Plugin {
                 }
                 format!("{}/{}", self.path, conf.plugin.main)
             } else {
-                self.lang.default_install(&self.path, &self.name).await?
+                self.lang.default_install(&self.path, &self.name, &verbose).await?
```
Comment (vincenzopalazzo) [inline coffee_lib/src/plugin.rs:127]:
```suggestion
                self.lang.default_install(&self.path, &self.name, verbose).await?
```

File: coffee_lib/src/plugin.rs [hunk 9]
```diff
@@ -107,10 +124,10 @@ impl Plugin {
...
             }
         } else {
-            self.lang.default_install(&self.path, &self.name).await?
+            self.lang.default_install(&self.path, &self.name, &verbose).await?
```
Comment (vincenzopalazzo) [inline coffee_lib/src/plugin.rs:130]:
```suggestion
            self.lang.default_install(&self.path, &self.name, verbose).await?
```

File: coffee_lib/src/plugin_manager.rs [hunk 10]
```diff
@@ -11,7 +11,7 @@ pub trait PluginManager {
     async fn configure(&mut self) -> Result<(), CoffeeError>;
 
     /// install a plugin by name, return an error if some error happens.
-    async fn install(&mut self, plugins: &str) -> Result<(), CoffeeError>;
+    async fn install(&mut self, plugins: &str, verbose: &bool) -> Result<(), CoffeeError>;
```
Comment (vincenzopalazzo) [inline coffee_lib/src/plugin_manager.rs:14]:
```suggestion
    async fn install(&mut self, plugins: &str, verbose: bool) -> Result<(), CoffeeError>;
```

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
This PR do what is expected but I think we can improve the code a little bit

**Comment by vincenzopalazzo** [comment]
Ah also the same change is required also  inside the `Plugin::configure` method inside the `if` branch

**Comment by Harshit933** [comment]
@vincenzopalazzo Thank you reviewing the PR. I'll update the code!

File: coffee_lib/src/plugin.rs (hunk 6 above)
Comment (Harshit933) [reply coffee_lib/src/plugin.rs:52]: Hmm I get it. After reading the docs, the output function would be better to debug if any errors occur while installing the plugins, whereas using the flag `-q` won't offer any other utilities for debugging.

File: coffee_cmd/src/main.rs [hunk 11]
```diff
@@ -11,11 +10,14 @@ use coffee_lib::plugin_manager::PluginManager;
 
 #[tokio::main]
 async fn main() -> Result<(), CoffeeError> {
+
```
Comment (vincenzopalazzo) [inline coffee_cmd/src/main.rs:13]: space no needed

File: coffee_lib/src/plugin.rs [hunk 12]
```diff
@@ -18,21 +19,26 @@ pub enum PluginLang {
...
+                    let _ = cmd
+                        .spawn()
+                        .expect("Unable to run the command")
+                        .wait()
+                        .await?;
+                } else {
+                    let _ = cmd.output().await?;
+                };
+
+
```
Comment (vincenzopalazzo) [inline coffee_lib/src/plugin.rs:41]: space no needed

**Review by vincenzopalazzo — COMMENTED** [review]
Hey looks good, just some spurious space to remove, also you need to fix the CI

If you are lucky I think running the `make` command could help you to fix it

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
The CI is still not happy https://github.com/coffee-tools/coffee/actions/runs/4293298443/jobs/7481181598

In addition, could you clear the commit? If I was you I was dividing the commitment as follow

- commit one: add the command line argument and the verbose parameter to the function
- commit two: implement the verbose logic

But if this it too complex you can also squash in a single one

**Comment by vincenzopalazzo** [comment]
@Harshit933 Having some problems in finishing this PR?

**Comment by Harshit933** [comment]
> @Harshit933 Having some problems in finishing this PR?

I'm still figuring it out how to organise the commit. I will try doing it today only

**Comment by vincenzopalazzo** [comment]
>I'm still figuring it out how to organize the commit. I will try doing it today only

Try to do this in a new branch :) otherwise, you will mess up the PR

In addition, if this is too difficult (and it is) you can just squash the commit in a single one. Usually the commit are already pushed organized and reworking it can be a little bit difficult

**Comment by Harshit933** [comment]
Can I simply create a new PR describing the outcomes of the code?

**Comment by Harshit933** [comment]
I think I've messed it up

**Comment by vincenzopalazzo** [comment]
You can now use the change to recreate the history that I propose you here https://github.com/coffee-tools/coffee/pull/51#pullrequestreview-1317719433

Just not open another PR with the same problem of the commits

**Comment by Harshit933** [comment]
Okay I understand, should I just push the code from that branch to the main one?

**Comment by vincenzopalazzo** [comment]
>Okay I understand, should I just push the code from that branch to the main one?

I'm not sure if I understand, but at this point the PR is simple why not rewrite the code and adjust the commit history without a problem?

Again git is a power tool so you can do different thing in different way, it is really up to you find the way that works for you

**Pull request closed without merging**